compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -186,6 +192,104 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         false
     }
 
+    // For generic associated types (GATs) which implied 'static requirement
+    // from higher-ranked trait bounds (HRTB). Try to locate span of the trait
+    // and the span which bounded to the trait for adding 'static lifetime suggestion
+    fn suggest_static_lifetime_for_gat_from_hrtb(
+        &self,
+        diag: &mut DiagnosticBuilder<'_, ErrorGuaranteed>,
+        lower_bound: RegionVid,
+    ) {
+        let mut suggestions = vec![];
+        let hir = self.infcx.tcx.hir();
+
+        // find generic associated types in the given region 'lower_bound'
+        let gat_id_and_generics = self
+            .regioncx
+            .placeholders_contained_in(lower_bound)
+            .map(|placeholder| {
+                if let Some(id) = placeholder.name.get_id()
+                    && let Some(placeholder_id) = id.as_local()
+                    && let Some(generics_impl) =
+                    hir.get_parent(hir.local_def_id_to_hir_id(placeholder_id.clone())).generics()
+                {
+                    Some((placeholder_id, generics_impl))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
```

Comment:
Seeing that you're only using it once afterwards, I don't think you need to collect this first, right? It can remain an iterator. Collecting has a marginal perf cost, but we can avoid it with confidence thanks to the type ownership system :)

Reply:
Yes, agreed. The reason for keeping the code this way I thought was for more readability and ease of debugging.